### DISCUSSION TITLE
isValid is false but errors is empty

### DISCUSSION THREAD

**Discussion by wereHamster** [react-hook-form/react-hook-form#4316] [Q&A]
Under which circumstances does `isValid` return false but `errors` is an empty object? What else (other than errors) can cause `isValid` to remain false?

**Comment by bluebill1049** [comment]
when an input is not been validated. can you share a CSB?

**Reply by wereHamster** [reply]
That will be difficult, the existing code is quite involved.

I see another strange thing: `dirtyFields` includes fields which are no longer part of the form. My code adds and removes fields as the user steppes through the form. When the user advances to the second step, the fields from the first step are removed from the DOM (and I assumed they would be automatically unregistered from the react-hook-form internal state). But they remain visible in `dirtyFields`.

I removed all validation rules from the fields in the second step, and when I advance to the second step the form is valid (which is to be expected, no field is required), but as soon as I interact with one of the fields the form turns invalid. Why would interacting with a field which has no validation rules turn the form invalid?

After advancing to the second step:
![image](https://user-images.githubusercontent.com/34538/109467170-148ff480-7a6b-11eb-839f-901965d65046.png)

After interacting with a field:
![image](https://user-images.githubusercontent.com/34538/109467210-240f3d80-7a6b-11eb-9c60-61f152d45219.png)

(the type/url fields are not visible in the second step, yet they remain in `dirtyFields`)

**Reply by bluebill1049** [reply]
which version are you using?

**Reply by wereHamster** [reply]
react-hook-form@6.15.3

**Reply by bluebill1049** [reply]
can you reproduce in a simple csb, otherwise it's hard to investigate the issue.

**Reply by wereHamster** [reply]
Another weird thing: one of my fields is not appearing in `touched` even though I touched it (it even failed the validation a bit before):

![image](https://user-images.githubusercontent.com/34538/109470219-6e92b900-7a6f-11eb-8d36-c4f0cdab20ef.png)

**Reply by wereHamster** [reply]
I'll try…

**Reply by wereHamster** [reply]
@bluebill1049 not sure if related, but found this: dirtyFields is nonEmpty and isDirty is false.

https://codesandbox.io/s/react-hook-form-isdirty-vs-dirtyfields-gfsc9?file=/src/index.js

Press the «Next» button.

**Reply by bluebill1049** [reply]
aha, that's because you don't have a `defaultValues`.

https://react-hook-form.com/api#formState

>Make sure to provide all inputs' defaultValues at the useForm, so hook form can have a single source of truth to compare whether the form is dirty.

**Reply by wereHamster** [reply]
I'm setting `defaultValue` in each `useController`, isn't that the same?

**Reply by bluebill1049** [reply]
should set defaultValues at useForm and you can skip those inline too.

**Reply by wereHamster** [reply]
The documentation of the `Controller` component says this to the `defaultValue` prop:

> You need to either set defaultValue at the field-level or call useForm with defaultValues.

But this is evidently not true, if the form behaves differently when I use it as opposed to passing `defaultValues` to `useForm`. Is that a bug or just inaccurate documentation?

**Reply by bluebill1049** [reply]
For dirty `formState`, always needs to set `defautlvalues` whether is register or controller, I would say the doc should be improved on that dot point.
